Code/Tools/FBuild/FBuildCore/Helpers/VSProjectGenerator.cpp

```diff
@@ -137,17 +224,51 @@ const AString & VSProjectGenerator::GenerateVCXProj( const AString & projectFile
                     break;
                 }
             }
+
+            const bool   isHeader = IsFilenameAHeaderFile( *fIt );
+            const char * includeTypeString = isHeader ? "ClInclude" : "CustomBuild";
+            Write( "    <%s Include=\"%s\">\n", includeTypeString, fileName );
             if ( fileType )
             {
-                Write( "    <CustomBuild Include=\"%s\">\n", fileName );
                 Write( "        <FileType>%s</FileType>\n", fileType );
-                Write( "    </CustomBuild>\n" );
             }
-            else
+
+            if ( writeCustomBuildCommands && !isHeader )
             {
-                Write( "    <CustomBuild Include=\"%s\" />\n", fileName );
+                // Build Command [Conditional]
+                Write( "        <Message>Compiling %%(Identity)</Message>\n" );
+                const VSProjectConfig * const cEnd = configs.End( );
+                for ( const VSProjectConfig * cIt = configs.Begin( ); cIt != cEnd; ++cIt )
+                {
+                    // If we have a .PlatformIncludePattern filter defined, and this platform doesn't match it, then skip emitting this platform to the .vcxproj
+                    if( ShouldWeSkipThisPlatform( cIt->m_PlatformIncludePattern, cIt->m_Platform ) )
+                    {
+                        continue;
+                    }
+
+                    const ObjectListNode * oln = ProjectGeneratorBase::FindTargetForIntellisenseInfo( cIt->m_Target );
+                    if ( oln != nullptr )
+                    {
+                        AString my_compilerName( ( oln != nullptr ) ? oln->GetCompiler( )->GetName( ).Get( ) : "error: compiler unknown" );
+                        AString my_compilerOptions( ( oln != nullptr ) ? oln->GetCompilerOptions( ).Get( ) : "error: compiler options unknown" );
+
+                        my_compilerOptions.Replace( "%1", "%(FullPath)" );
+                        my_compilerOptions.Replace( "%2", "$(OutDir)%(Identity).obj" );
+
+                        Write( "        <Command Condition=\"'$(Configuration)|$(Platform)'=='%s|%s'\">@echo on &amp; SET %s; &amp; %s %s</Command>\n", cIt->m_Config.Get( ), GetPossiblyAliasedPlatformStringInUppercase( cIt->m_TargetArchitectureAlias, cIt->m_Platform ), environment, my_compilerName.Get( ), my_compilerOptions.Get( ) );
```

Comment:
environment is a double-nul terminated string, so this will only print the first entry in the environment list. Does this need to be processed somehow?